### ISSUE TITLE
Support and import and export of HTML code

### ISSUE THREAD

**Issue by slightfoot** [slightfoot/flutter_rich_text_editor#1]

**Comment by matthew-carroll** [comment]
@bk-one @salihgueler let's figure out where this ticket belongs in terms of priority

**Comment by bk-one** [comment]
not important right now

**Comment by hillelcoren** [comment]
Sorry to hear it, FWIW it's important to us...

**Comment by bk-one** [comment]
Let's aim then for the next release? We do only add markdown right now, but happy to reevaluate after we released 0.1 in the coming weeks

**Comment by hillelcoren** [comment]
That would be awesome, thank you!

**Comment by hillelcoren** [comment]
That said if there's support for markdown I believe it's backwards/forward convertible to HTML so it should work for us as is

**Comment by salihgueler** [comment]
@hillelcoren please check `serializeDocumentToMarkdown` and `deserializeMarkdownToDocument` functions for Markdown implementation.

**Comment by matthew-carroll** [comment]
@hillelcoren have you or anyone on your team tried out the editor? I was generally aware of your interest in the editor but wasn't aware that you were keeping an eye on it. Can you say a little bit about your timeline for adoption and list your MVP features for the editor?

**Comment by hillelcoren** [comment]
I've actually jus started working on it, if the editor is ready we'd go live with it immediately. We were going to use HTML but if markdown is ready that could work for us, our main use case is to generate HTML based emails.

**Comment by matthew-carroll** [comment]
We won't know if it's ready until you tell us it's ready :)

I would recommend trying to integrate it and then filing issues for any bugs you come across, or MVP feature requests. As you file issue, it would also be helpful to see any visual designs or screenshots of that part of your product in action so that we can glean surrounding context.

**Comment by hillelcoren** [comment]
I spent a few hours this morning working on it, here are my notes.

- I had trouble getting `pub get` to complete on the stable channel but following a suggestion from @matthew-carroll I was able to get it to work on dev.
- The wiki is very useful, small problem I noticed is the links in the top section of [this page](https://github.com/superlistapp/super_editor/wiki) are broken
- I haven't been able to try the widget due to the error below
- From reading the docs I noticed this: "An Editor does not include any app bar controls or popup controls". Will controls be added. I realize I could build them myself I guess with some Flutter widgets but it would be nice if it came with controls out of the box.
- For our app it's important all platforms are supported (web, mobile and desktop)

```
super_editor/lib/src/default_editor/document_interaction.dart:639:31: Error: A value of type 'bool' 
can't be returned from a function with return type 'KeyEventResult'.
 - 'KeyEventResult' is from 'package:flutter/src/widgets/focus_manager.dart' 
 ('../flutter/packages/flutter/lib/src/widgets/focus_manager.dart').
      onKey: (node, event) => true,
                              ^
```

Happy to create a separate issue if needed...

**Comment by matthew-carroll** [comment]
@hillelcoren I think a separate issue would be good. Let's keep this issue tracking HTML support.

When you file a new issue, can you include the code that causes the error, and include the platform you're running on and the `flutter --version` info?

**Comment by matthew-carroll** [comment]
@hillelcoren can you outline your requirements for HTML (de)serialization? It looks like this ticket was filed as an empty placeholder, so we'll need to figure out what the specific deliverable is before we can work on this.

**Comment by hillelcoren** [comment]
Our main use case initially is to provide a WYSIWYG editor for [HTML emails](https://en.wikipedia.org/wiki/HTML_email). We'd like to support basic formatting (bold, italic, font size, etc) but don't need to support complex layouts. Note: markdown would also work.

In the future we'd like to support something closer to this: https://grapesjs.com

Thanks for your work on the project, it's already pretty impressive!

**Comment by matthew-carroll** [comment]
@hillelcoren if markdown will work for you then that would make more sense because we already support markdown.

grapejs looks to be a full-blown application that goes far beyond document layouts. You may be able to use some of the building blocks from `super_editor` in building such a tool, but general layout definitely goes beyond the goals of document layout and editing.

Until we hear otherwise from a customer, I'm going to consider this ticket as being on-hold because there doesn't appear to be anyone that truly needs this behavior at this time.

**Comment by sabetAI** [comment]
Hi @matthew-carroll, for my use-case I'm parsing and pre-processing html documents, and currently rendering them in a `ListView` with [HtmlWidgets](https://pub.dev/packages/flutter_widget_from_html). `HtmlWidget` currently doesn't support selectable text, so it would be convenient if I can render directly using `super_editor`.

**Comment by matthew-carroll** [comment]
Hi @sabetAI. I have a couple questions about your use-case:

- what subset of the HTML spec do you need to be supported?
- is it possible for you to use `super_editor` by leveraging an existing package to go from HTML to markdown, and then from markdown back to HTML?

**Comment by sabetAI** [comment]
Hi @matthew-carroll, 
- it's mostly just individual `<p>` and `<a>` elements at the moment, with some styling ie bold font, italics, sup. Nothing layout-related. 
- since the html is just individual elements with no nesting, perhaps I can pre-process with some html-to-markdown library (ie [htm2md](https://pub.dev/packages/html2md) on frontend, or something like [html2markdown](https://pypi.org/project/html2markdown/) on the backend).

**Comment by matthew-carroll** [comment]
@sabetAI yeah, my guess is that you can do HTML -> markdown without issue. Please post back after you get a chance to try that and let us know if it works for your needs.

I can't remember for sure, but I think the `markdown` package that we're already including may offer HTML-to-markdown translation.

**Comment by britannio** [comment]
I've got a basic implementation with test coverage, it's probably worth completing https://github.com/superlistapp/super_editor/issues/86 before integrating it in the repo though as it depends on the [html](https://pub.dev/packages/html) package.
![image](https://user-images.githubusercontent.com/33752528/127325981-f2afab7f-349f-47c4-aca2-169c30149099.png)

**Comment by bugrevealingbme** [comment]
I think this issue should be explained better. How can I keep this document suitable for api. SuperEditor export not explained

**Comment by matthew-carroll** [comment]
@bugrevealingbme can you please rephrase your comment as a specific question?
